### PR TITLE
v1.0.3

### LINKED ISSUE #9: why would you use a language extension?
https://github.com/es3n1n/portal2-internal/blob/823110d3ee743a8561cbb4f1b2e427ae1cbe4731/portal2-internal/hack/cfg/cfg.cpp#L22

---
Doesn't the __forceinline here suck?

Should we inline this in the first place, even? I assume this is going to destroy the optimizations of clang/msvc once we get a lot more config variables.

We/you probably should avoid compiler extensions.

Good work, big man.

### LINKED ISSUE #12: a fun todo list
https://github.com/es3n1n/portal2-internal/blob/823110d3ee743a8561cbb4f1b2e427ae1cbe4731/portal2-internal/sdk/misc/color_t.hpp#L49-L61

This should be a truth table, with the values assigned at the top. A temporary color_t construction would be clearer, imo:
```cpp
[[nodiscard]] color_t from_hsb(float h, float s, float b) {
   color_t tmp{};
   //
   //
   //: do stuff..
   
   switch(h):
      case 1:
         tmp.r = ...;
         tmp.b = ...;
         tmp.g = ...;
         return tmp;
      case 2:
         tmp.r = ...;
         tmp.b = ...;
         tmp.g = ...;
         return tmp;
```
et cetera.

- [x] done / fixed / invalid

---

https://github.com/es3n1n/portal2-internal/blob/823110d3ee743a8561cbb4f1b2e427ae1cbe4731/portal2-internal/sdk/misc/color_t.hpp#L6-L6

Those should be uint8_t. So should be the parameters in member functions.

- [x] done / fixed / invalid

---

https://github.com/es3n1n/portal2-internal/blob/823110d3ee743a8561cbb4f1b2e427ae1cbe4731/portal2-internal/sdk/misc/color_t.hpp#L19-L21

This is highly unsafe. This probably should just error out on shit data.

- [x] done / fixed / invalid

---

https://github.com/es3n1n/portal2-internal/blob/823110d3ee743a8561cbb4f1b2e427ae1cbe4731/portal2-internal/sdk/misc/color_t.hpp#L25-L25

I suggest this:
```cpp
    std::uint32_t get_hex() {
        auto get_hex_byte = [&](std::uint8_t field, unsigned shift_amt) -> std::uint8_t {
            return (field & 0xff) << shift_amt;
        };
        
        enum {
            ALPHA = 24,
            RED = 16,
            GREEN = 8,
            BLUE = 0
        };

        return  get_hex_byte(this->a, ALPHA) |
                get_hex_byte(this->r, RED) |
                get_hex_byte(this->g, GREEN) |
                get_hex_byte(this->b, BLUE);
    }
```

- [x] done / fixed / invalid

---

Throughout the project you use `#pragma once` directives.

These were not supported by GCC. (Yeah, it was a long time ago. Poor little GNU compiler). Still, the `#ifndef MY_SHIT` is highly preferrable.

- [x] done / fixed / invalid

---

https://github.com/es3n1n/portal2-internal/blob/823110d3ee743a8561cbb4f1b2e427ae1cbe4731/portal2-internal/sdk/misc/client_class.hpp#L6-L7

- Why not `using create_client_class_fn`? more modern shit, looks cool, LSP's love them.

- Should these be publicly available?

- [x] done / fixed / invalid

---

https://github.com/es3n1n/portal2-internal/blob/823110d3ee743a8561cbb4f1b2e427ae1cbe4731/portal2-internal/sdk/misc/vec_t.hpp#L11

Clang optimizes these away. This just restricts the usage availability.

- [x] done / fixed / invalid

---

There's probably a lot more that needs to be done, but these are the only ones that i've found.

### PR DESCRIPTION
**Pull request by es3n1n**

### CODE REVIEW

**Comment by cpz** [comment]
image
https://github.com/es3n1n/portal2-internal/pull/14/files#diff-fbc99da07453ccea1d84f1463718c54ed075f661ad439e4a07a8f6b5649766eeR10

Нельзя придумать чот типо std::array<std::tuple<color, fmt>>? Конструкция какая-нибудь (враппер), чтоб везде его юзать.

Colors PushColors(std::array<std::tuple<color, fmt>> sorry_im_being_autis)

**Comment by cpz** [comment]
https://github.com/es3n1n/portal2-internal/pull/14/files#diff-a9ca7c3a276efc0c0ef7d47d128d8ec7beb6a7be39a7dbe70d9e5a9664a7f498R38

Сейм, враппер бы, чот типо 
```
template <typename Where, typename T>
void push(t* ptr, std::string_view name)
{
if conx
where.pushback
}
```

**Comment by kehrazy** [comment]
> image
> https://github.com/es3n1n/portal2-internal/pull/14/files#diff-fbc99da07453ccea1d84f1463718c54ed075f661ad439e4a07a8f6b5649766eeR10
> 
> Нельзя придумать чот типо std::array<std::tuple<color, fmt>>? Конструкция какая-нибудь (враппер), чтоб везде его юзать.
> 
> Colors PushColors(std::array<std::tuple<color, fmt>> sorry_im_being_autis)

std::array плохо работает с мсвц (premium compiletime, idk what to say)..

if that ain't a concern - how would the api look like?

std::find is shit

isn't a hashmap better here?

**Comment by kehrazy** [comment]
idk about inlining a json based operation ever.. its fucking json, it will always be black magic

the color_t union looks neat as fuck

**Comment by kehrazy** [comment]
also, we can use compiletime black magic to get our format strings, the usage would be like jm's xorstr:

```cpp
"{0}"_format(...)
```

although im not so sure if we should. is std format fast?

**Comment by es3n1n** [comment]
> Нельзя придумать чот типо std::array<std::tuple<color, fmt>>? Конструкция какая-нибудь (враппер), чтоб везде его юзать.
> 
> Colors PushColors(std::array<std::tuple<color, fmt>> sorry_im_being_autis)

sounds reasonable 

> std::array плохо работает с мсвц (premium compiletime, idk what to say)..

we can use initializer_list(if we aren't building it at the run time) or unordered_map instead

> if that ain't a concern - how would the api look like?

at this point im not really sure how would the api for manipulating with current cfg structs would look like.. we(read me) should recode the whole config system anyway so maybe we can keep things simple for now

> std::find is shit

why so?

> isn't a hashmap better here?

exactly

**Comment by es3n1n** [comment]
> idk about inlining a json based operation ever.. its fucking json, it will always be black magic

we're inlining it because since this function has template parameters, for every possible parameter combination it will create a unique function in the .text section. i'm not sure that it makes any sense if a function is inlined or not, either way, it would be a mess in the produced file. but if we are inlining it at least the output would be smaller since we aren't doing prologues/epilogues

> the color_t union looks neat as fuck

haha thanks


> also, we can use compiletime black magic to get our format strings, the usage would be like jm's xorstr:
> 
> ```c++
> "{0}"_format(...)
> ```
> 
> although im not so sure if we should. is std format fast?

std::format is fast enough, although having a cutom string literal for formatting strings sounds cool

**Comment by kehrazy** [comment]
> > std::find is shit
> 
> why so?

a bidirectional for loop instead of a jump table... idk, man

**Comment by es3n1n** [comment]
> > > std::find is shit
> > 
> > 
> > why so?
> 
> a bidirectional for loop instead of a jump table... idk, man

fair enough

**Comment by es3n1n** [comment]
also @kehrazy 
i just found out that in your code if i pass `__int64` (or `time_t`) as a template parameter to `util::detail::is_safe_castable` concept, it throws errors because `std::numeric_limits<__int64>::is_iec559` is set to false.
this is why i had to change the time_t cast back to static_cast.
it doesn't really seem like the correct behaviour to me

i think we are almost ready for merging it to master tho. i'd be glad if someone can review the entire PR

**Review by cpz — APPROVED** [review]
[Approved](https://www.youtube.com/watch?v=rjvImL06_1k). Пока вы лежите на диване, они делают лучший портал2 хак!!!!

**Review by mostdefinitelynotnth — APPROVED** [review]
просто невероятно! филигранная работа! именно так выглядит продукт мастера своего ремесла.

**Comment by kehrazy** [comment]
> also @kehrazy 
> i just found out that in your code if i pass `__int64` (or `time_t`) as a template parameter to `util::detail::is_safe_castable` concept, it throws errors because `std::numeric_limits<__int64>::is_iec559` is set to false.
> this is why i had to change the time_t cast back to static_cast.
> it doesn't really seem like the correct behaviour to me

yeah, that was expected.

the whole `safe_cast` ordeal was to avoid ieee754 shenanigans while casting... we could probably do some template matching there - but is it really clear? the naming probably sucks, it should be named `float_cast` or something

> 
> i think we are almost ready for merging it to master tho. i'd be glad if someone can review the entire PR

good shit!

File: portal2-internal/hack/cfg/opts.hpp [hunk 1]
```diff
@@ -16,6 +16,10 @@ namespace opts {
         color_t m_portal_2 = {100, 255, 255, 255};
 
         color_t& operator[](size_t index) {
+            return at(index);
+        }
+
+        color_t& at(size_t index) {
             switch (index) {
             case 0:
```
Comment (kehrazy) [inline portal2-internal/hack/cfg/opts.hpp]: kinda hate how portal 1 is at the index 0..

hb a `enum portal` with literally portal1 and portal2

**Comment by kehrazy** [comment]
other than that, lgtm

**Review by kehrazy — APPROVED** [review]

**Review by kehrazy — APPROVED** [review]

File: portal2-internal/hack/cfg/opts.hpp (hunk 1 above)
Comment (es3n1n) [reply portal2-internal/hack/cfg/opts.hpp]: we treat the portal option struct as an array where indexation starts at 0, but yeah sure we can also add an enum with said indices

**Comment by es3n1n** [comment]
> the whole `safe_cast` ordeal was to avoid ieee754 shenanigans while casting... we could probably do some template matching there - but is it really clear? the naming probably sucks, it should be named `float_cast` or something

maybe we can just use the original name `clamp_cast`

**Comment by kehrazy** [comment]
> > the whole `safe_cast` ordeal was to avoid ieee754 shenanigans while casting... we could probably do some template matching there - but is it really clear? the naming probably sucks, it should be named `float_cast` or something
> 
> maybe we can just the original name `clamp_cast`

sure

**Review by kehrazy — APPROVED** [review]
lgtm

**Pull request merged**